File: Second Time/C++/src/5_Longest_Palindromic_Substring/5_Longest_Palindromic_Substring.cpp
/*
 * @Author: fieryheart 
 * @Date: 2019-11-23 13:51:19 
 * @Last Modified by: fiery
 * @Last Modified time: 2019-11-23 15:14:34
 * @Desc: the key of solving the problem is to take account into the repeat number of center part of palindromic substring.
 * Firstly remove the repeat number of center part, then check whether two sides of subtring is equal.
 */

#include "../../lib/Root.h"

//==============================================================

class Solution {
public:
    string longestPalindrome(string s) {
        if(s.length() < 2) return s;
        int maxl = 0, start = 0;
        for(int i = 0; s.length() - i > maxl / 2; ) {
            int left, right;
            left = right = i;
            while(right+1 < s.length() && s[right+1] == s[right]) right++;
            i = right+1;
            while(right+1 < s.length() && left > 0 && s[left-1] == s[right+1]) {
                left--;
                right++;
            }
            if(right-left+1 > maxl) {
                maxl = right-left+1;
                start = left;
            }
            
        }
        return s.substr(start, maxl);        
    }
};

//==============================================================

int main()
{
    string s = "babad";
    Solution solu;
    string ans = solu.longestPalindrome(s);
    Print::print(ans);
    return 0;
}

File: Second Time/Python/5.Longest-Palindromic-Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        res = ""
        for i in range(len(s)):
            tmp = self.findPalindrome(s, i, i)
            if len(tmp) > len(res):
                res = tmp
            tmp = self.findPalindrome(s, i, i+1)
            if len(tmp) > len(res):
                res = tmp
        return res
    def findPalindrome(self, s, left, right) -> str:
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return s[left+1:right]